[c10] make intrusive_ptr available as a pybind holder type

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42944 [c10d][jit] initial torchbind bindings for ProcessGroup
* #47075 [c10d] switch ProcessGroupNCCL:Options to be managed by intrusive_ptr
* #47074 [c10d] switch Store to be managed by intrusive_ptr
* #44046 [c10d] switch ProcessGroup::Work to be managed by intrusive_ptr
* **#44492 [c10] make intrusive_ptr available as a pybind holder type**



Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

gmagogsfm commented on c10/util/intrusive_ptr.h:
```diff
@@ -207,14 +207,24 @@ class intrusive_ptr final {
   // pointers except from inside the make_intrusive() and
   // weak_intrusive_ptr::lock() implementations
-  explicit intrusive_ptr(TTarget* target) noexcept : target_(target) {}
+  explicit intrusive_ptr(TTarget* target, bool add_ref=true) noexcept : target_(target) {
```
Why is it needed to inc ref here?

gmagogsfm commented on c10/util/intrusive_ptr.h:
```diff
@@ -207,15 +207,28 @@ class intrusive_ptr final {
+  //    to be a part of public API
+  //    https://pybind11.readthedocs.io/en/stable/advanced/smart_ptrs.html#custom-smart-pointers
+  explicit intrusive_ptr(TTarget* target, bool add_ref=true) noexcept : target_(target) {
+    if (target_ != NullType::singleton() && add_ref) {
```
It makes sense to not refcount NullType::singleton. Just curious, anything in the code that prevents refcounting in it?

wanchaol replied:
iirc all the increfs inside intrusive_ptr is guarded with the check if the `target_ != NullType::singleton()`, otherwise runtime it will get memory corruption as it does not have a refcount

gmagogsfm reviewed (approved):
Change looks good. But this is a somewhat dangerous change, please make sure all tests are passing. Right now there are some failures.

smessmer commented on c10/util/intrusive_ptr.h:
```diff
@@ -207,15 +207,28 @@ class intrusive_ptr final {
+  // 2. Pybind11 custom smart pointer requires pointer/raw pointer constructor
+  //    to be a part of public API
+  //    https://pybind11.readthedocs.io/en/stable/advanced/smart_ptrs.html#custom-smart-pointers
+  explicit intrusive_ptr(TTarget* target, bool add_ref=true) noexcept : target_(target) {
```
I would rather not make this public API. Using the regular intrusive_ptr APIs should be safe and you shouldn't be able to shoot yourself in the foot with them. We have the `raw::incref()` APIs for people who know what they're doing that opt in to dangerous behavior by using the `raw` namespace.

Also, defaulting this to true disables the protection we have against making intrusive_ptrs to stack objects. It is intentional that the intrusive_ptr does not increment refcount but the `make_intrusive` does it. See the comment about "Stack allocated intrusive_ptr_target safety" at the top of the file.

wanchaol committed:
Update on "[c10] make intrusive_ptr available as a pybind holder type"


Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] make intrusive_ptr available as a pybind holder type"


Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

[ghstack-poisoned]

gmagogsfm reviewed (approved):
Awesome!

wanchaol committed:
Update on "[c10] make intrusive_ptr available as a pybind holder type"


Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] make intrusive_ptr available as a pybind holder type"


Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] make intrusive_ptr available as a pybind holder type"


Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] make intrusive_ptr available as a pybind holder type"


Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] make intrusive_ptr available as a pybind holder type"


Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] make intrusive_ptr available as a pybind holder type"


Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] make intrusive_ptr available as a pybind holder type"


Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] make intrusive_ptr available as a pybind holder type"


Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] make intrusive_ptr available as a pybind holder type"


Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] make intrusive_ptr available as a pybind holder type"


Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] make intrusive_ptr available as a pybind holder type"


Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] make intrusive_ptr available as a pybind holder type"


Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

[ghstack-poisoned]

smessmer commented on c10/util/intrusive_ptr.h:
```diff
@@ -347,19 +378,12 @@ class intrusive_ptr final {
-    ++result.target_->weakcount_;
-
-    return result;
+    return intrusive_ptr(new TTarget(std::forward<Args>(args)...));
```
You can probably delete `intrusive_ptr::make()` and instead call the constructor directly from `make_intrusive`.

smessmer reviewed (approved):
Looks good, thanks.

wanchaol committed:
Update on "[c10] make intrusive_ptr available as a pybind holder type"


Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] make intrusive_ptr available as a pybind holder type"


Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] make intrusive_ptr available as a pybind holder type"


Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] make intrusive_ptr available as a pybind holder type"




Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] make intrusive_ptr available as a pybind holder type"




Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10] make intrusive_ptr available as a pybind holder type"




Differential Revision: [D23632278](https://our.internmc.facebook.com/intern/diff/D23632278)

[ghstack-poisoned]

zhangguanheng66 commented:
This PR breaks our building blocks in torchtext. To reproduce the error, please build pytorch/torchtext from source code and run
```
from torchtext.experimental.transforms import (
    PRETRAINED_SP_MODEL,
    sentencepiece_processor,
)
from torchtext.utils import download_from_url

sp_model_path = download_from_url(PRETRAINED_SP_MODEL['text_bpe_25000'])
spm_transform = sentencepiece_processor(sp_model_path)
jit_spm_transform = torch.jit.script(spm_transform.to_ivalue())
```

The error message
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-7-4aebdbcd2a5c> in <module>()
      7 sp_model_path = download_from_url(PRETRAINED_SP_MODEL['text_bpe_25000'])
      8 spm_transform = sentencepiece_processor(sp_model_path)
----> 9 jit_spm_transform = torch.jit.script(spm_transform.to_ivalue())

4 frames
/usr/local/lib/python3.6/dist-packages/torch/jit/_script.py in script(obj, optimize, _frames_up, _rcb)
    896     if isinstance(obj, torch.nn.Module):
    897         return torch.jit._recursive.create_script_module(
--> 898             obj, torch.jit._recursive.infer_methods_to_compile
    899         )
    900 

/usr/local/lib/python3.6/dist-packages/torch/jit/_recursive.py in create_script_module(nn_module, stubs_fn, share_types)
    368     check_module_initialized(nn_module)
    369     concrete_type = get_module_concrete_type(nn_module, share_types)
--> 370     return create_script_module_impl(nn_module, concrete_type, stubs_fn)
    371 
    372 def create_script_module_impl(nn_module, concrete_type, stubs_fn):

/usr/local/lib/python3.6/dist-packages/torch/jit/_recursive.py in create_script_module_impl(nn_module, concrete_type, stubs_fn)
    424 
    425     # Actually create the ScriptModule, initializing it with the function we just defined
--> 426     script_module = torch.jit.RecursiveScriptModule._construct(cpp_module, init_fn)
    427 
    428     # Compile methods if necessary

/usr/local/lib/python3.6/dist-packages/torch/jit/_script.py in _construct(cpp_module, init_fn)
    386             """
    387             script_module = RecursiveScriptModule(cpp_module)
--> 388             init_fn(script_module)
    389 
    390             # Finalize the ScriptModule: replace the nn.Module state with our

/usr/local/lib/python3.6/dist-packages/torch/jit/_recursive.py in init_fn(script_module)
    389             orig_value = getattr(nn_module, name)
    390             orig_value = orig_value.value if isinstance(orig_value, torch.jit.Attribute) else orig_value
--> 391             cpp_module.setattr(name, orig_value)
    392 
    393         # 2. Copy the submodules from the original `nn_module` to the new ScriptModule,

RuntimeError: Unable to cast Python instance to C++ type (compile in debug mode for details)
```
